code/components/extra-natives-five/src/ColshapeNatives.cpp

```diff
@@ -0,0 +1,1086 @@
+		{
+			trace("CREATE_COLSHAPE_POLYZONE: Too many vertices (%zu, max is %zu)\n",
+			obj.via.array.size, MAX_VERTICES);
+			context.SetResult(false);
+			return;
+		}
+
+		std::vector<Vector2> points;
+		points.reserve(obj.via.array.size);
+
+		for (size_t i = 0; i < obj.via.array.size; ++i)
+		{
+			msgpack::object& element = obj.via.array.ptr[i];
+			float x = 0.0f, y = 0.0f;
+
+			// makes it possible to have arrays in lua tables (if for some reason you want to do that, mostly you'd use x = 1, y = 2 or vector2(1,2) etc..)
+			if (element.type == msgpack::type::ARRAY)
+			{
+				// If array of [x,y] or [x,y,z]
+				auto& arr = element.via.array;
+				if (arr.size >= 2)
+				{
+					// x, y
+					x = arr.ptr[0].as<float>();
+					y = arr.ptr[1].as<float>();
+					// if arr.size == 3 => there's a z, but we ignore it
+				}
+				else
+				{
+					trace("CREATE_COLSHAPE_POLYZONE: Array element at %zu has <2 floats.\n", i);
+					context.SetResult(false);
+					return;
+				}
+			}
+			// this is essentially { x = 1.0, y= 2.0 }
+			else if (element.type == msgpack::type::MAP)
+			{
+				for (size_t j = 0; j < element.via.map.size; ++j)
+				{
+					auto& kv = element.via.map.ptr[j];
+					if (kv.key.type == msgpack::type::STR)
+					{
+						std::string key(kv.key.via.str.ptr, kv.key.via.str.size);
+						if (key == "x")
+							x = kv.val.as<float>();
+						if (key == "y")
+							y = kv.val.as<float>();
+					}
+				}
+			}
+			else if (element.type == msgpack::type::EXT)
+			{
+
+				int8_t extType = element.via.ext.type();
+				size_t dataSize = element.via.ext.size;
+				const char* dataPtr = element.via.ext.ptr;
+				
+				// just to make sure its vector2, vector3 or vector4
+				// vector2 is 20, found in MsgPackSerializer.cs in the clrcore project (packer.PackExtendedTypeValue(20, ms.ToArray());)
+				// vector3 is 21, vector 4 is 22 etc..
+				if (dataSize <= 16 && dataSize >= 8 && ((extType == 20 && dataSize == 8) || (extType == 21 && dataSize == 12) || (extType == 22 && dataSize == 16)))
+				{
+					dataPtr++; // skip the first byte as it contains information about type/size which we don't need here
+					float x_val, y_val;
+					memcpy(&x_val, dataPtr, sizeof(float));
+					memcpy(&y_val, dataPtr + 4, sizeof(float));
+					x = x_val;
+					y = y_val;
+				}
+				else
+				{
+					trace("CREATE_COLSHAPE_POLYZONE: Unexpected EXT type or size at %zu\n", i);
+					context.SetResult(false);
+					return;
+				}
+			}
+			else
+			{
+				trace("CREATE_COLSHAPE_POLYZONE: Unexpected element type at index %zu\n", i);
+				context.SetResult(false);
+				return;
+			}
+
+			points.push_back(Vector2(x, y));
+		}
+		bool success = ColShapeManager::Get().CreatePolyzone(colShapeId, points, minZ, maxZ);
+		context.SetResult<bool>(success);
+	});
+
+
+	// use game thread instead of seperate thread as suggested by Heron
+	static auto lastUpdateTime = std::chrono::high_resolution_clock::now();
+	OnMainGameFrame.Connect([]()
```

Comment:
What do you think about a method to enable the colshape manager? Could be a native or a ConVar. This would also allow adjusting the hardcoded bounds depending on the requirements of the server that is using it.